internal/observability/graph.go

```diff
@@ -71,25 +71,106 @@ type graph struct {
 	meter              *metrics.Scope
 	logger             *zap.Logger
 	extract            ContextExtractor
-	metricTagsBlockMap map[string]struct{}
+	metricTagsBlockMap *metricsTagIgnore
 
 	edgesMu sync.RWMutex
 	edges   map[string]*edge
 
 	inboundLevels, outboundLevels levels
 }
 
-func newGraph(meter *metrics.Scope, logger *zap.Logger, extract ContextExtractor, metricTagsBlocklist []string) graph {
-	metricTagsBlockMap := make(map[string]struct{}, len(metricTagsBlocklist))
+// if the field is set to true, the metrics tag won't be emitted
+// it is a more performant way than have a map (field lookup versus map lookup)
+type metricsTagIgnore struct {
+	source          bool
+	dest            bool
+	transport       bool
+	procedure       bool
+	encoding        bool
+	routingKey      bool
+	routingDelegate bool
+	direction       bool
+	rpcType         bool
+}
+
+func newMetricsTagIgnore(metricTagsBlocklist []string) *metricsTagIgnore {
+	r := new(metricsTagIgnore)
 	for _, m := range metricTagsBlocklist {
-		metricTagsBlockMap[m] = struct{}{}
+		switch m {
+		case _source:
+			r.source = true
+		case _dest:
+			r.dest = true
+		case _transport:
+			r.transport = true
+		case _procedure:
+			r.procedure = true
+		case _encoding:
+			r.encoding = true
+		case _routingKey:
+			r.routingKey = true
+		case _routingDelegate:
+			r.routingDelegate = true
+		case _direction:
+			r.direction = true
+		case _rpcType:
+			r.rpcType = true
+		}
+	}
+	return r
+}
+
+func (m *metricsTagIgnore) tags(req *transport.Request, direction string, rpcType transport.Type) metrics.Tags {
+	tags := metrics.Tags{
+		_source:          req.Caller,
+		_dest:            req.Service,
+		_transport:       unknownIfEmpty(req.Transport),
+		_procedure:       req.Procedure,
+		_encoding:        string(req.Encoding),
+		_routingKey:      req.RoutingKey,
+		_routingDelegate: req.RoutingDelegate,
+		_direction:       direction,
+		_rpcType:         rpcType.String(),
+	}
+
+	if !m.source {
```

Comment:
Should this be `if m.source` (here and below) based on the comment on the struct and its ctor?

This is overwriting "real" values with dropped values, so this isn't a "should this be emitted" check, it's a "should it be dropped" check - while `!m.source` should mean "source is allowed".

At least, I think that's the intent - maybe I'm misunderstanding?

Reply:
ohh yea good call. And I was missing unit test on those :(